Add `download_retry_count` for `direct` and `network_mirror` in CLI configuration

yottta opened the pull request:
<!--

** Thank you for your contribution! Please read this carefully! **

Please make sure you go through the checklist below. If your PR does not meet all requirements, please file it
as a draft PR. Core team members will only review your PR once it meets all the requirements below (unless your
change is something as trivial as a typo fix).

-->

<!-- If your PR resolves an issue, please add it here. -->
Resolves #3186
Follow up on #3338

Add configuration in the CLI configuration for being able to granularly control the max retry count for each provider source. For now, only the `direct` and `network_mirror` methods support this new attribute. 

For the `oci_mirror`, to support such configuration we should rethink the way the underlying http client is built since right now it supports none of the known configuration for other sources (TF_REGISTRY_DISCOVERY_RETRY,  TF_REGISTRY_CLIENT_TIMEOUT, TF_PROVIDER_DOWNLOAD_RETRY).


## Checklist

<!-- Please check of ALL items in this list for all PRs: -->

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.
- [x] I (and other contributors to this PR) have not looked at the Terraform source code while implementing this PR.

### Go checklist

<!-- If your PR contains Go code, please make sure you check off all items on this list: --> 

- [x] I have run golangci-lint on my change and receive no errors relevant to my code.
- [x] I have run existing tests to ensure my code doesn't break anything.
- [x] I have added tests for all relevant use cases of my code, and those tests are passing.
- [x] I have only exported functions, variables and structs that should be used from other packages.
- [x] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

<!-- If you have changed the website, please follow this checklist: -->

- [x] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

lbordowitz commented on website/docs/cli/config/config-file.mdx:
```diff
@@ -278,6 +278,29 @@ identity, but a network mirror with a TLS certificate can potentially serve
+You can use this argument to disable fully the retries in case of a specific method (`download_retry_count = 0`).
+:::info
+When used together with [`TF_PROVIDER_DOWNLOAD_RETRY`](./environment-variables.mdx#tf_provider_download_retry), the configuration value
+will be used for each method that is configured with. For the methods
```
Nit: "that it is configured with"

yottta commented on internal/command/e2etest/provider_network_mirror_test.go:
```diff
@@ -0,0 +1,148 @@
+			envVars:                  nil,
+			expectedErrMsg:           "/example.com/test/test/terraform-provider-test_0.0.1_linux_amd64.zip giving up after 2 attempt(s)",
+		},
+		"defined tofurc.network_mirror.download_retry_count as 1, TF_PROVIDER_DOWNLOAD_RETRY defined as 2, tofurc used": {
```
Good question.
I agree that this is in contradiction with the newly added ones in #3256. 

The reason I did it this way, was to ensure that the current behaviour is unchanged, meaning that the env var will work properly as it did till now, and will be applied to all the methods configured in the CLI configuration but with no `download_retry_count` argument.
The idea is to be able to use the env var for all methods and then overwrite it with the CLI configuration for each method that it's needed to.
If we swap the precedence, then if all the methods are configured, the env var will overwrite all of those.

Maybe, for the sake of consistency, we should swap the precedence 🤔

yottta commented on internal/command/cliconfig/provider_installation.go:
```diff
@@ -564,6 +577,7 @@ type ProviderInstallationMethod struct {
 	Location ProviderInstallationLocation
 	Include  []string `hcl:"include"`
 	Exclude  []string `hcl:"exclude"`
```
New PR submitted to remove this: https://github.com/opentofu/opentofu/pull/3391
Let's merge that only after we merge this one to have clear history in `main`.

yottta commented:
Realised that I introduced the same configuration for the `oci_mirror` without actually using it. Removed it in https://github.com/opentofu/opentofu/pull/3368/commits/01b598997916bb9f89a4c5fcfd803b7b30952c2e
Also created another issue to handle these settings for `oci_mirror` separately: https://github.com/opentofu/opentofu/issues/3392
